File: src/shmif/arcan_shmif_mousestate.c
#include <arcan_shmif.h>
#include <pthread.h>
#include "platform/shmif_platform.h"
#include "shmif_privint.h"

void arcan_shmif_mousestate_setup(
	struct arcan_shmif_cont* con, int flags, uint8_t* state)
{
	if (!con || !con->priv)
		return;

	struct mstate* ms = (struct mstate*) state;
	if (!ms)
		ms = &con->priv->mstate;

	*ms = (struct mstate){
		.rel = (flags & (~ARCAN_MOUSESTATE_NOCLAMP)) > 0,
		.noclamp = !!(flags & ARCAN_MOUSESTATE_NOCLAMP)
	};
}

static bool absclamp(
	struct mstate* ms, struct arcan_shmif_cont* con,
	int* out_x, int* out_y)
{
	if (ms->ax < 0 && !ms->noclamp)
		ms->ax = 0;
	else
		ms->ax = ms->ax > con->w && !ms->noclamp ? con->w : ms->ax;

	if (ms->ay < 0 && !ms->noclamp)
		ms->ay = 0;
	else
		ms->ay = ms->ay > con->h && !ms->noclamp ? con->h : ms->ay;

/* with clamping, we can get relative samples that shouldn't
 * propagate, so test that before updating history */
	bool res = ms->ly != ms->ay || ms->lx != ms->ax;
	*out_y = ms->ly = ms->ay;
	*out_x = ms->lx = ms->ax;

	return res;
}

bool arcan_shmif_mousestate(
	struct arcan_shmif_cont* con, uint8_t* state,
	struct arcan_event* inev, int* out_x, int* out_y)
{
	return arcan_shmif_mousestate_ioev(con, state, &inev->io, out_x, out_y);
}

/*
 * Weak attempt of trying to bring some order in the accumulated mouse
 * event handling chaos - definitely one of the bigger design fails that
 * can't be fixed easily due to legacy.
 */
bool arcan_shmif_mousestate_ioev(
	struct arcan_shmif_cont* con, uint8_t* state,
	struct arcan_ioevent* inev, int* out_x, int* out_y)
{
	struct mstate* ms = (struct mstate*) state;
	if (!con || !con->priv)
		return false;

	if (!state)
		ms = &con->priv->mstate;

	if (!ms|| !out_x || !out_y)
		return false;

	if (!inev){
		if (!ms->inrel)
			return absclamp(ms, con, out_x, out_y);
		else
			*out_x = *out_y = 0;
		return true;
	}

	if (!ms ||
		inev->datatype != EVENT_IDATATYPE_ANALOG ||
		inev->devkind != EVENT_IDEVKIND_MOUSE
	)
		return false;

/* state switched between samples, reset tracking */
	bool gotrel = inev->input.analog.gotrel;
	if (gotrel != ms->inrel){
		ms->inrel = gotrel;
		ms->ax = ms->ay = ms->lx = ms->ly = 0;
	}

/* packed, both axes in one sample */
	if (inev->subid == 2){
/* relative input sample, are we in relative state? */
		if (gotrel){
/* good case, the sample is already what we want */
			if (ms->rel){
				*out_x = ms->lx = inev->input.analog.axisval[0];
				*out_y = ms->ly = inev->input.analog.axisval[2];
				return *out_x || *out_y;
			}
/* bad case, the sample is relative and we want absolute,
 * accumulate and clamp */
			ms->ax += inev->input.analog.axisval[0];
			ms->ay += inev->input.analog.axisval[2];

			return absclamp(ms, con, out_x, out_y);
		}
/* good case, the sample is absolute and we want absolute, clamp */
    else {
			if (!ms->rel){
				ms->ax = inev->input.analog.axisval[0];
				ms->ay = inev->input.analog.axisval[2];
				return absclamp(ms, con, out_x, out_y);
			}
/* worst case, the sample is absolute and we want relative,
 * need history AND discard large jumps */
			int dx = inev->input.analog.axisval[0] - ms->lx;
			int dy = inev->input.analog.axisval[2] - ms->ly;
			ms->lx = inev->input.analog.axisval[0];
			ms->ly = inev->input.analog.axisval[2];
			if (!dx && !dy){
				return false;
			}
			*out_x = dx;
			*out_y = dy;
			return true;
		}
	}

/* one sample, X axis */
	else if (inev->subid == 0){
		if (gotrel){
			if (ms->rel){
				*out_x = ms->lx = inev->input.analog.axisval[0];
				return *out_x;
			}
			ms->ax += inev->input.analog.axisval[0];
			return absclamp(ms, con, out_x, out_y);
		}
		else {
			if (!ms->rel){
				ms->ax = inev->input.analog.axisval[0];
				return absclamp(ms, con, out_x, out_y);
			}
			int dx = inev->input.analog.axisval[0] - ms->lx;
			ms->lx = inev->input.analog.axisval[0];
			if (!dx)
				return false;
			*out_x = dx;
			*out_y = 0;
			return true;
		}
	}

/* one sample, Y axis */
	else if (inev->subid == 1){
		if (gotrel){
			if (ms->rel){
				*out_y = ms->ly = inev->input.analog.axisval[0];
				return *out_y;
			}
			ms->ay += inev->input.analog.axisval[0];
			return absclamp(ms, con, out_x, out_y);
		}
		else {
			if (!ms->rel){
				ms->ay = inev->input.analog.axisval[0];
				return absclamp(ms, con, out_x, out_y);
			}
			int dy = inev->input.analog.axisval[0] - ms->ly;
			ms->ly = inev->input.analog.axisval[0];
			if (!dy)
				return false;
			*out_x = 0;
			*out_y = dy;
			return true;
		}
	}
	else
		return false;
}

File: src/shmif/arcan_shmif_mousestate.zig
// Zig reimplementation of arcan_shmif_mousestate.c
// Drop-in C-ABI-compatible replacement for mousestate functions.
//
// Exports: arcan_shmif_mousestate_setup, arcan_shmif_mousestate,
//          arcan_shmif_mousestate_ioev
//
const std = @import("std");
const builtin = @import("builtin");
const is_freestanding = (builtin.os.tag == .freestanding);
const off = @import("shmif_offsets");
const c = @import("shmif_types");

export fn arcan_shmif_mousestate_setup(
    con: ?*c.struct_arcan_shmif_cont,
    flags: c_int,
    state: ?[*]u8,
) void {
    if (is_freestanding) return;
    if (con == null) return;
    const con_v = con.?;
    if (con_v.priv == null) return;

    var ms: *anyopaque = undefined;
    if (state) |s| {
        ms = @ptrCast(@alignCast(s));
    } else {
        ms = off.Hidden.getMstatePtr(@ptrCast(@alignCast(con_v.priv)));
    }

    off.Mstate.zeroMstate(ms);
    const noclamp_mask: c_int = c.ARCAN_MOUSESTATE_NOCLAMP;
    off.Mstate.setRel(ms, @intFromBool((flags & ~noclamp_mask) > 0));
    off.Mstate.setNoclamp(ms, @intFromBool((flags & noclamp_mask) != 0));
}

/// Clamp absolute coordinates and update history.
/// Returns true if position changed since last sample.
fn absclamp(
    ms: *anyopaque,
    con: *c.struct_arcan_shmif_cont,
    out_x: *c_int,
    out_y: *c_int,
) bool {
    const cw: i32 = @intCast(con.w);
    const ch: i32 = @intCast(con.h);

    if (off.Mstate.getAx(ms) < 0 and off.Mstate.getNoclamp(ms) == 0)
        off.Mstate.setAx(ms, 0)
    else if (off.Mstate.getAx(ms) > cw and off.Mstate.getNoclamp(ms) == 0)
        off.Mstate.setAx(ms, cw);

    if (off.Mstate.getAy(ms) < 0 and off.Mstate.getNoclamp(ms) == 0)
        off.Mstate.setAy(ms, 0)
    else if (off.Mstate.getAy(ms) > ch and off.Mstate.getNoclamp(ms) == 0)
        off.Mstate.setAy(ms, ch);

    // with clamping, we can get relative samples that shouldn't
    // propagate, so test that before updating history
    const res = off.Mstate.getLy(ms) != off.Mstate.getAy(ms) or off.Mstate.getLx(ms) != off.Mstate.getAx(ms);
    off.Mstate.setLy(ms, off.Mstate.getAy(ms));
    out_y.* = off.Mstate.getLy(ms);
    off.Mstate.setLx(ms, off.Mstate.getAx(ms));
    out_x.* = off.Mstate.getLx(ms);

    return res;
}

export fn arcan_shmif_mousestate(
    con: ?*c.struct_arcan_shmif_cont,
    state: ?[*]u8,
    inev: ?*c.arcan_event,
    out_x: ?*c_int,
    out_y: ?*c_int,
) bool {
    if (is_freestanding) return false;
    if (inev) |ev| {
        return arcan_shmif_mousestate_ioev(con, state, ev.io(), out_x, out_y);
    }
    return arcan_shmif_mousestate_ioev(con, state, null, out_x, out_y);
}

// Weak attempt of trying to bring some order in the accumulated mouse
// event handling chaos - definitely one of the bigger design fails that
// can't be fixed easily due to legacy.
export fn arcan_shmif_mousestate_ioev(
    con: ?*c.struct_arcan_shmif_cont,
    state: ?[*]u8,
    inev: ?*c.arcan_ioevent,
    out_x: ?*c_int,
    out_y: ?*c_int,
) bool {
    if (is_freestanding) return false;
    if (con == null) return false;
    const con_v = con.?;
    if (con_v.priv == null) return false;

    var ms: *anyopaque = undefined;
    if (state) |s| {
        ms = @ptrCast(@alignCast(s));
    } else {
        ms = off.Hidden.getMstatePtr(@ptrCast(@alignCast(con_v.priv)));
    }

    const ox = out_x orelse return false;
    const oy = out_y orelse return false;

    if (inev == null) {
        if (off.Mstate.getInrel(ms) == 0)
            return absclamp(ms, con_v, ox, oy)
        else {
            ox.* = 0;
            oy.* = 0;
        }
        return true;
    }

    const ev = inev.?;

    if (ev.datatype != c.EVENT_IDATATYPE_ANALOG or
        ev.devkind != c.EVENT_IDEVKIND_MOUSE)
        return false;

    // state switched between samples, reset tracking
    const gotrel: u8 = if (ev.input.analog.gotrel != 0) 1 else 0;
    if (gotrel != off.Mstate.getInrel(ms)) {
        off.Mstate.setInrel(ms, gotrel);
        off.Mstate.setAx(ms, 0);
        off.Mstate.setAy(ms, 0);
        off.Mstate.setLx(ms, 0);
        off.Mstate.setLy(ms, 0);
    }

    const subid = ev.unnamed_0.unnamed_0.subid;

    // packed, both axes in one sample
    if (subid == 2) {
        if (gotrel != 0) {
            // relative input sample
            if (off.Mstate.getRel(ms) != 0) {
                // good case, the sample is already what we want
                off.Mstate.setLx(ms,ev.input.analog.axisval[0]);
                ox.* = off.Mstate.getLx(ms);
                off.Mstate.setLy(ms,ev.input.analog.axisval[2]);
                oy.* = off.Mstate.getLy(ms);
                return ox.* != 0 or oy.* != 0;
            }
            // bad case, the sample is relative and we want absolute,
            // accumulate and clamp
            off.Mstate.setAx(ms,off.Mstate.getAx(ms) + ev.input.analog.axisval[0]);
            off.Mstate.setAy(ms,off.Mstate.getAy(ms) + ev.input.analog.axisval[2]);
            return absclamp(ms, con_v, ox, oy);
        } else {
            // good case, the sample is absolute and we want absolute, clamp
            if (off.Mstate.getRel(ms) == 0) {
                off.Mstate.setAx(ms,ev.input.analog.axisval[0]);
                off.Mstate.setAy(ms,ev.input.analog.axisval[2]);
                return absclamp(ms, con_v, ox, oy);
            }
            // worst case, the sample is absolute and we want relative,
            // need history AND discard large jumps
            const dx = ev.input.analog.axisval[0] - off.Mstate.getLx(ms);
            const dy = ev.input.analog.axisval[2] - off.Mstate.getLy(ms);
            off.Mstate.setLx(ms,ev.input.analog.axisval[0]);
            off.Mstate.setLy(ms,ev.input.analog.axisval[2]);
            if (dx == 0 and dy == 0) {
                return false;
            }
            ox.* = dx;
            oy.* = dy;
            return true;
        }
    }

    // one sample, X axis
    else if (subid == 0) {
        if (gotrel != 0) {
            if (off.Mstate.getRel(ms) != 0) {
                off.Mstate.setLx(ms,ev.input.analog.axisval[0]);
                ox.* = off.Mstate.getLx(ms);
                return ox.* != 0;
            }
            off.Mstate.setAx(ms,off.Mstate.getAx(ms) + ev.input.analog.axisval[0]);
            return absclamp(ms, con_v, ox, oy);
        } else {
            if (off.Mstate.getRel(ms) == 0) {
                off.Mstate.setAx(ms,ev.input.analog.axisval[0]);
                return absclamp(ms, con_v, ox, oy);
            }
            const dx = ev.input.analog.axisval[0] - off.Mstate.getLx(ms);
            off.Mstate.setLx(ms,ev.input.analog.axisval[0]);
            if (dx == 0)
                return false;
            ox.* = dx;
            oy.* = 0;
            return true;
        }
    }

    // one sample, Y axis
    else if (subid == 1) {
        if (gotrel != 0) {
            if (off.Mstate.getRel(ms) != 0) {
                off.Mstate.setLy(ms,ev.input.analog.axisval[0]);
                oy.* = off.Mstate.getLy(ms);
                return oy.* != 0;
            }
            off.Mstate.setAy(ms,off.Mstate.getAy(ms) + ev.input.analog.axisval[0]);
            return absclamp(ms, con_v, ox, oy);
        } else {
            if (off.Mstate.getRel(ms) == 0) {
                off.Mstate.setAy(ms,ev.input.analog.axisval[0]);
                return absclamp(ms, con_v, ox, oy);
            }
            const dy = ev.input.analog.axisval[0] - off.Mstate.getLy(ms);
            off.Mstate.setLy(ms,ev.input.analog.axisval[0]);
            if (dy == 0)
                return false;
            ox.* = 0;
            oy.* = dy;
            return true;
        }
    } else {
        return false;
    }
}
